From fbb58d9c25aa112164a3deeeb18a869b9e67832c Mon Sep 17 00:00:00 2001
From: vrochalo <vinicius.rochalobo@windriver.com>
Date: Tue, 25 Nov 2025 11:14:27 -0300
Subject: [PATCH] Update glance store config

This patch syncs the Glance backend configuration with
the updated model from OpenStack-Helm. The upstream
version referenced by Armada still included deprecated
backend fields, which were removed or updated in later
commits (884fdc05c94b, ff1582dd3a92, c4c80371f9a3).

Their relevant changes were consolidated into this single
patch, keeping only what is needed to update the backend
configuration and enable the newer backend workflow
(including NetApp integration).

Signed-off-by: vrochalo <Vinicius.RochaLobo@windriver.com>
---
 glance/Chart.yaml                      |  2 +-
 glance/templates/configmap-etc.yaml    | 30 ++++++++++++--------------
 glance/templates/deployment-api.yaml   | 17 +++++++++------
 glance/templates/job-storage-init.yaml | 19 +++++-----------
 glance/templates/pvc-images.yaml       |  2 +-
 glance/values.yaml                     | 27 +++++++++++++++++++++++
 glance/values_overrides/tls.yaml       |  1 +
 releasenotes/notes/glance.yaml         |  4 ++++
 8 files changed, 64 insertions(+), 38 deletions(-)

diff --git a/glance/Chart.yaml b/glance/Chart.yaml
index c5adfefa..549138c8 100644
--- a/glance/Chart.yaml
+++ b/glance/Chart.yaml
@@ -14,7 +14,7 @@ apiVersion: v1
 appVersion: v1.0.0
 description: OpenStack-Helm Glance
 name: glance
-version: 0.4.23
+version: 0.5.0
 home: https://docs.openstack.org/glance/latest/
 icon: https://www.openstack.org/themes/openstack/images/project-mascots/Glance/OpenStack_Project_Glance_vertical.png
 sources:
diff --git a/glance/templates/configmap-etc.yaml b/glance/templates/configmap-etc.yaml
index 210560b0..70dfd040 100644
--- a/glance/templates/configmap-etc.yaml
+++ b/glance/templates/configmap-etc.yaml
@@ -68,27 +68,25 @@ limitations under the License.
 {{- $_ := tuple "image" "public" "api" . | include "helm-toolkit.endpoints.keystone_endpoint_uri_lookup" | set .Values.conf.glance.DEFAULT "public_endpoint" -}}
 {{- end -}}
 
-{{- if empty .Values.conf.glance.glance_store.stores -}}
+{{- if empty .Values.conf.glance.DEFAULT.enabled_backends -}}
 {{- if eq .Values.storage "rbd" }}
-{{- $_ := "file, http, rbd" | set .Values.conf.glance.glance_store "stores" -}}
-{{- end -}}
-{{- if eq .Values.storage "pvc" }}
-{{- $_ := "file, http" | set .Values.conf.glance.glance_store "stores" -}}
-{{- end -}}
-{{ if or (eq .Values.storage "radosgw") (eq .Values.storage "swift") }}
-{{- $_ := "file, http, swift" | set .Values.conf.glance.glance_store "stores" -}}
+{{- $_ := "file:file,http:http,rbd:rbd" | set .Values.conf.glance.DEFAULT "enabled_backends" -}}
+{{- else if or (eq .Values.storage "radosgw") (eq .Values.storage "swift") }}
+{{- $_ := "file:file,http:http,swift:swift" | set .Values.conf.glance.DEFAULT "enabled_backends" -}}
+{{- else -}}
+{{/* pvc, local (hostPath) or other/ephemeral (emptyDir) */}}
+{{- $_ := "file:file,http:http" | set .Values.conf.glance.DEFAULT "enabled_backends" -}}
 {{- end -}}
 {{- end -}}
 
-{{- if empty .Values.conf.glance.glance_store.default_store -}}
+{{- if empty .Values.conf.glance.glance_store.default_backend -}}
 {{- if eq .Values.storage "rbd" }}
-{{- $_ := "rbd" | set .Values.conf.glance.glance_store "default_store" -}}
-{{- end -}}
-{{- if eq .Values.storage "pvc" }}
-{{- $_ := "file" | set .Values.conf.glance.glance_store "default_store" -}}
-{{- end -}}
-{{ if or (eq .Values.storage "radosgw") (eq .Values.storage "swift") }}
-{{- $_ := "swift" | set .Values.conf.glance.glance_store "default_store" -}}
+{{- $_ := "rbd" | set .Values.conf.glance.glance_store "default_backend" -}}
+{{- else if or (eq .Values.storage "radosgw") (eq .Values.storage "swift") }}
+{{- $_ := "swift" | set .Values.conf.glance.glance_store "default_backend" -}}
+{{- else -}}
+{{/* pvc, local (hostPath) or other/ephemeral (emptyDir) */}}
+{{- $_ := "file" | set .Values.conf.glance.glance_store "default_backend" -}}
 {{- end -}}
 {{- end -}}
 
diff --git a/glance/templates/deployment-api.yaml b/glance/templates/deployment-api.yaml
index 3ba5f17a..4cb910a6 100644
--- a/glance/templates/deployment-api.yaml
+++ b/glance/templates/deployment-api.yaml
@@ -89,19 +89,19 @@ spec:
             - chown
             - -R
             - "glance:"
-            - {{ .Values.conf.glance.glance_store.filesystem_store_datadir }}
+            - {{ .Values.conf.glance.file.filesystem_store_datadir }}
           volumeMounts:
             - name: pod-tmp
               mountPath: /tmp
             - name: glance-images
-              mountPath: {{ .Values.conf.glance.glance_store.filesystem_store_datadir }}
+              mountPath: {{ .Values.conf.glance.file.filesystem_store_datadir }}
         {{ if eq .Values.storage "rbd" }}
         - name: ceph-keyring-placement
 {{ tuple $envAll "glance_api" | include "helm-toolkit.snippets.image" | indent 10 }}
 {{ dict "envAll" $envAll "application" "glance" "container" "ceph_keyring_placement" | include "helm-toolkit.snippets.kubernetes_container_security_context" | indent 10 }}
           env:
             - name: RBD_STORE_USER
-              value: {{ .Values.conf.glance.glance_store.rbd_store_user | quote }}
+              value: {{ .Values.conf.glance.rbd.rbd_store_user | quote }}
           command:
             - /tmp/ceph-keyring.sh
           volumeMounts:
@@ -209,11 +209,11 @@ spec:
             {{- end }}
             {{- end }}
             - name: glance-etc
-              mountPath: {{ .Values.conf.glance.glance_store.swift_store_config_file }}
+              mountPath: {{ .Values.conf.glance.swift.swift_store_config_file }}
               subPath: swift-store.conf
               readOnly: true
             - name: glance-images
-              mountPath: {{ .Values.conf.glance.glance_store.filesystem_store_datadir }}
+              mountPath: {{ .Values.conf.glance.file.filesystem_store_datadir }}
 {{- if eq .Values.storage "cinder" }}
             - name: host-rootfs
               mountPath: /mnt/host-rootfs
@@ -287,7 +287,12 @@ spec:
         - name: glance-images
           persistentVolumeClaim:
             claimName: glance-images
-{{ else }}
+{{ else if eq .Values.storage "local" }}
+        - name: glance-images
+          hostPath:
+            path: {{ .Values.conf.glance.file.filesystem_store_datadir }}
+            type: DirectoryOrCreate
+{{- else }}
         - name: glance-images
           emptyDir: {}
 {{- end }}
diff --git a/glance/templates/job-storage-init.yaml b/glance/templates/job-storage-init.yaml
index 22cd4109..5e6a210a 100644
--- a/glance/templates/job-storage-init.yaml
+++ b/glance/templates/job-storage-init.yaml
@@ -114,17 +114,17 @@ spec:
               value: {{ .Values.storage | quote }}
             {{- if eq .Values.storage "rbd" }}
             - name: RBD_POOL_NAME
-              value: {{ .Values.conf.glance.glance_store.rbd_store_pool | quote }}
+              value: {{ .Values.conf.glance.rbd.rbd_store_pool | quote }}
             - name: RBD_POOL_APP_NAME
               value: {{ .Values.conf.software.rbd.rbd_store_pool_app_name | quote }}
             - name: RBD_POOL_USER
-              value: {{ .Values.conf.glance.glance_store.rbd_store_user | quote }}
+              value: {{ .Values.conf.glance.rbd.rbd_store_user | quote }}
             - name: RBD_POOL_REPLICATION
-              value: {{ .Values.conf.glance.glance_store.rbd_store_replication | quote }}
+              value: {{ .Values.conf.glance.rbd.rbd_store_replication | quote }}
             - name: RBD_POOL_CRUSH_RULE
-              value: {{ .Values.conf.glance.glance_store.rbd_store_crush_rule | quote }}
+              value: {{ .Values.conf.glance.rbd.rbd_store_crush_rule | quote }}
             - name: RBD_POOL_CHUNK_SIZE
-              value: {{ .Values.conf.glance.glance_store.chunk_size | quote }}
+              value: {{ .Values.conf.glance.rbd.chunk_size | quote }}
             - name: RBD_POOL_SECRET
               value: {{ .Values.secrets.rbd | quote }}
             {{ end }}
@@ -166,10 +166,6 @@ spec:
               readOnly: true
             {{ end }}
             {{ end }}
-            {{- if eq .Values.storage "pvc" }}
-            - name: glance-images
-              mountPath: {{ .Values.conf.glance.glance_store.filesystem_store_datadir }}
-            {{ end }}
 {{- dict "enabled" .Values.manifests.certificates "name" .Values.secrets.tls.image.api.public | include "helm-toolkit.snippets.tls_volume_mount" | indent 12 }}
       volumes:
         - name: pod-tmp
@@ -191,10 +187,5 @@ spec:
             secretName: {{ .Values.ceph_client.user_secret_name }}
         {{ end }}
         {{ end }}
-        {{- if eq .Values.storage "pvc" }}
-        - name: glance-images
-          persistentVolumeClaim:
-            claimName: glance-images
-        {{ end }}
 {{- dict "enabled" .Values.manifests.certificates "name" .Values.secrets.tls.image.api.public | include "helm-toolkit.snippets.tls_volume" | indent 8 }}
 {{- end }}
diff --git a/glance/templates/pvc-images.yaml b/glance/templates/pvc-images.yaml
index 21ea9613..9451a959 100644
--- a/glance/templates/pvc-images.yaml
+++ b/glance/templates/pvc-images.yaml
@@ -25,7 +25,7 @@ metadata:
     helm.sh/resource-policy: keep
   {{- end }}
 spec:
-  accessModes: [ "ReadWriteOnce" ]
+  accessModes: {{ .Values.volume.accessModes }}
   resources:
     requests:
       storage: {{ .Values.volume.size }}
diff --git a/glance/values.yaml b/glance/values.yaml
index c902bcab..bad40e5c 100644
--- a/glance/values.yaml
+++ b/glance/values.yaml
@@ -263,6 +263,10 @@ conf:
       memcache_security_strategy: ENCRYPT
       service_type: image
     glance_store:
+      # Since 2024.1 this section must contain the only key 'default_backend'.
+      # Other keys should be defined in the corresponding per-backend sections.
+      # This is for backward compatibility.
+      filesystem_store_datadir: /var/lib/glance/images
       cinder_catalog_info: volumev3::internalURL
       chunk_size: 8
       rbd_store_replication: 3
@@ -270,7 +274,28 @@ conf:
       rbd_store_pool: glance.images
       rbd_store_user: glance
       rbd_store_ceph_conf: /etc/ceph/ceph.conf
+      default_swift_reference: ref1
+      swift_store_container: glance
+      swift_store_create_container_on_put: true
+      swift_store_config_file: /etc/glance/swift-store.conf
+      swift_store_endpoint_type: internalURL
+    file:
       filesystem_store_datadir: /var/lib/glance/images
+    # These two sections os_glance_tasks_store and os_glance_staging_store
+    # are mandatory. Glance will be unable to delete images from if these
+    # two are not properly configured.
+    os_glance_tasks_store:
+      filesystem_store_datadir: /var/lib/glance/tmp/os_glance_tasks_store
+    os_glance_staging_store:
+      filesystem_store_datadir: /var/lib/glance/tmp/os_glance_staging_store
+    rbd:
+      rbd_store_chunk_size: 8
+      rbd_store_replication: 3
+      rbd_store_crush_rule: replicated_rule
+      rbd_store_pool: glance.images
+      rbd_store_user: glance
+      rbd_store_ceph_conf: /etc/ceph/ceph.conf
+    swift:
       default_swift_reference: ref1
       swift_store_container: glance
       swift_store_create_container_on_put: true
@@ -427,6 +452,8 @@ network:
 volume:
   class_name: general
   size: 2Gi
+  accessModes:
+    - ReadWriteOnce
 
 dependencies:
   dynamic:
diff --git a/glance/values_overrides/tls.yaml b/glance/values_overrides/tls.yaml
index d1aa35ad..86cba608 100644
--- a/glance/values_overrides/tls.yaml
+++ b/glance/values_overrides/tls.yaml
@@ -7,6 +7,7 @@ conf:
     keystone_authtoken:
       cafile: /etc/glance/certs/ca.crt
     glance_store:
+      # This option has been removed in 2024.1
       https_ca_certificates_file: /etc/glance/certs/ca.crt
       swift_store_cacert: /etc/glance/certs/ca.crt
     oslo_messaging_rabbit:
diff --git a/releasenotes/notes/glance.yaml b/releasenotes/notes/glance.yaml
index 3456e15c..f3438339 100644
--- a/releasenotes/notes/glance.yaml
+++ b/releasenotes/notes/glance.yaml
@@ -57,4 +57,8 @@ glance:
   - 0.4.21 Use uWSGI
   - 0.4.22 Enable custom annotations for Openstack secrets
   - 0.4.23 Update images used by default
+  - 0.4.24 Do not attach backend pvc to storage init pod
+  - 0.4.25 Allow customisation of pvc storage accessMode so we can run multiple api pods
+  - 0.4.26 Use quay.io/airshipit/kubernetes-entrypoint:latest-ubuntu_focal by default
+  - 0.5.0 Remove deprecated config options `stores` and `default_store`
 ...
-- 
2.34.1

